### ISSUE TITLE
java.lang.NoSuchMethodError: com.fasterxml.jackson.core.util.BufferRecycler.releaseToPool()V

### ISSUE THREAD

**Issue by ndepomereu** [sashirestela/simple-openai#89]
Hi,

For your users, and a future release fix: I had this error in my app at a ChatGPT call, with no stack trace:
`java.lang.NoSuchMethodError: com.fasterxml.jackson.core.util.BufferRecycler.releaseToPool()V`

I googled and found:
https://stackoverflow.com/questions/27746750/exception-in-thread-main-java-lang-nosuchmethoderror-com-fasterxml-jackson-co

I did what was said in best accepted answer, aka explicitly add to my pom.xml:

```
    <dependency>
        <groupId>com.fasterxml.jackson.core</groupId>
        <artifactId>jackson-annotations</artifactId>
        <version>2.17.0</version>
    </dependency>

    <dependency>
        <groupId>com.fasterxml.jackson.core</groupId>
        <artifactId>jackson-core</artifactId>
        <version>2.17.0</version>
    </dependency>
```

(Version 2.17.0 is the one used in simple-openai 2.2.0.)

Then my app worked back again.

Best,
Nicolas

**Comment by sashirestela** [comment]
@ndepomereu How can I reproduce the issue, please?

**Comment by ndepomereu** [comment]
Sorry, it was my app that would hide the stack trace.

This sample example triggers the Exception:

	SimpleOpenAI openai = SimpleOpenAI.builder()
		    .apiKey(ParmsUtil.getChatGptKey())
		    .build();
	

	ChatRequest chatRequest = ChatRequest.builder()
		    .model("gpt-4-turbo-preview")
		    .messages(List.of(
		        new ChatMsgSystem("You are an expert in AI."),
		        new ChatMsgUser("Write a technical article about ChatGPT, no more than 100 words.")))
		    .temperature(0.0)
		    .maxTokens(200)
		    .build();
	
	CompletableFuture<ChatResponse> futureChat = openai.chatCompletions().create(chatRequest);
	ChatResponse chatResponse = futureChat.join();
	System.out.println(chatResponse.firstContent());
```

```
Exception in thread "main" java.lang.NoSuchMethodError: com.fasterxml.jackson.core.util.BufferRecycler.releaseToPool()V
	at com.fasterxml.jackson.databind.ObjectMapper.writeValueAsString(ObjectMapper.java:4047)
	at io.github.sashirestela.cleverclient.util.JsonUtil.objectToJson(JsonUtil.java:38)
	at io.github.sashirestela.cleverclient.http.HttpProcessor.getBodyObject(HttpProcessor.java:140)
	at io.github.sashirestela.cleverclient.http.HttpProcessor.resolve(HttpProcessor.java:106)
	at io.github.sashirestela.cleverclient.http.HttpProcessor.invoke(HttpProcessor.java:79)
	at com.sun.proxy.$Proxy7.createRoot(Unknown Source)
	at io.github.sashirestela.openai.OpenAI$ChatCompletions.create(OpenAI.java:162)
	at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:710)
	at io.github.sashirestela.cleverclient.http.HttpProcessor.invoke(HttpProcessor.java:77)
	at com.sun.proxy.$Proxy7.create(Unknown Source)
	at com.chatmotorapi.samples.Chatbot.main(Chatbot.java:35)
```

It's probably a dependency conflict, here is my pom.xml:
https://api.agilotext.com/misc/pom.xml

If I uncomments the jackson dependencies, it works fine again.

Best,
Nicolas

**Comment by ndepomereu** [comment]
Sorry, it was my app that would hide the stack trace.

This sample example triggers the Exception:

	SimpleOpenAI openai = SimpleOpenAI.builder()
		    .apiKey(ParmsUtil.getChatGptKey())
		    .build();
	

	ChatRequest chatRequest = ChatRequest.builder()
		    .model("gpt-4-turbo-preview")
		    .messages(List.of(
		        new ChatMsgSystem("You are an expert in AI."),
		        new ChatMsgUser("Write a technical article about ChatGPT, no more than 100 words.")))
		    .temperature(0.0)
		    .maxTokens(200)
		    .build();
	
	CompletableFuture<ChatResponse> futureChat = openai.chatCompletions().create(chatRequest);
	ChatResponse chatResponse = futureChat.join();
	System.out.println(chatResponse.firstContent());
```


Exception in thread "main" java.lang.NoSuchMethodError: com.fasterxml.jackson.core.util.BufferRecycler.releaseToPool()V
	at com.fasterxml.jackson.databind.ObjectMapper.writeValueAsString(ObjectMapper.java:4047)
	at io.github.sashirestela.cleverclient.util.JsonUtil.objectToJson(JsonUtil.java:38)
	at io.github.sashirestela.cleverclient.http.HttpProcessor.getBodyObject(HttpProcessor.java:140)
	at io.github.sashirestela.cleverclient.http.HttpProcessor.resolve(HttpProcessor.java:106)
	at io.github.sashirestela.cleverclient.http.HttpProcessor.invoke(HttpProcessor.java:79)
	at com.sun.proxy.$Proxy7.createRoot(Unknown Source)
	at io.github.sashirestela.openai.OpenAI$ChatCompletions.create(OpenAI.java:162)
	at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:710)
	at io.github.sashirestela.cleverclient.http.HttpProcessor.invoke(HttpProcessor.java:77)
	at com.sun.proxy.$Proxy7.create(Unknown Source)
	at com.chatmotorapi.samples.Chatbot.main(Chatbot.java:35)
```

It's probably a dependency conflict, here is my pom.xml:
https://api.agilotext.com/misc/pom.xml

If I uncomments the jackson dependencies, it works fine again.

Best,
Nicolas

**Comment by sashirestela** [comment]
@ndepomereu I've tried your code several times and I cannot replicate the issue, the result is success always, as you can see in the below image:

![image](https://github.com/sashirestela/simple-openai/assets/14045506/7cbf8405-0e4b-4f3c-8ee2-e6f6b0b1833f)

I have run your code in an independent project with this pom file in two versions: with explicit Jackson dependencies and without them, and the result is the same.

```
<?xml version="1.0" encoding="UTF-8"?>

<project xmlns="http://maven.apache.org/POM/4.0.0"
  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
  <modelVersion>4.0.0</modelVersion>

  <groupId>io.github.sashirestela</groupId>
  <artifactId>simple-openai-playground</artifactId>
  <version>1.0.0</version>

  <name>simple-openai-playground</name>

  <properties>
    <!-- General -->
    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
    <maven.compiler.release>11</maven.compiler.release>
    <!-- Dependencies Versions -->
    <jackson.version>2.17.0</jackson.version>
    <simple-openai.version>2.1.0</simple-openai.version>
    <slf4j.version>2.0.12</slf4j.version>
    <!-- Plugins Versions -->
    <compiler.version>3.12.1</compiler.version>
    <exec.version>3.2.0</exec.version>
  </properties>

  <dependencyManagement>
    <dependencies>
      <dependency>
        <groupId>com.fasterxml.jackson.core</groupId>
        <artifactId>jackson-databind</artifactId>
        <version>${jackson.version}</version>
      </dependency>
      <dependency>
        <groupId>com.fasterxml.jackson.core</groupId>
        <artifactId>jackson-core</artifactId>
        <version>${jackson.version}</version>
      </dependency>
      <dependency>
        <groupId>io.github.sashirestela</groupId>
        <artifactId>simple-openai</artifactId>
        <version>${simple-openai.version}</version>
      </dependency>
      <dependency>
        <groupId>org.slf4j</groupId>
        <artifactId>slf4j-api</artifactId>
        <version>${slf4j.version}</version>
      </dependency>
      <dependency>
        <groupId>org.slf4j</groupId>
        <artifactId>slf4j-simple</artifactId>
        <version>${slf4j.version}</version>
        <optional>true</optional>
      </dependency>
    </dependencies>
  </dependencyManagement>

  <dependencies>
    <dependency>
      <groupId>com.fasterxml.jackson.core</groupId>
      <artifactId>jackson-databind</artifactId>
    </dependency>
    <dependency>
      <groupId>com.fasterxml.jackson.core</groupId>
      <artifactId>jackson-core</artifactId>
    </dependency>
    <dependency>
      <groupId>io.github.sashirestela</groupId>
      <artifactId>simple-openai</artifactId>
    </dependency>
    <dependency>
      <groupId>org.slf4j</groupId>
      <artifactId>slf4j-api</artifactId>
    </dependency>
    <dependency>
      <groupId>org.slf4j</groupId>
      <artifactId>slf4j-simple</artifactId>
    </dependency>
  </dependencies>

  <build>
    <plugins>
      <plugin>
        <groupId>org.apache.maven.plugins</groupId>
        <artifactId>maven-compiler-plugin</artifactId>
        <version>${compiler.version}</version>
      </plugin>
      <plugin>
        <groupId>org.codehaus.mojo</groupId>
        <artifactId>exec-maven-plugin</artifactId>
        <version>${exec.version}</version>
        <configuration>
          <cleanupDaemonThreads>false</cleanupDaemonThreads>
        </configuration>
      </plugin>
    </plugins>
  </build>
</project>
```

So, in summary, the issue is not related to simple-openai. If you don't have more observations, I think this issue could be closed.

**Comment by sashirestela** [comment]
Closing because it is not an issue.
